Update http-context.md (#22698)
dotnet/AspNetCore.Docs: aspnetcore/fundamentals/http-context.md

---
title: Access HttpContext in ASP.NET Core
author: coderandhiker

> description: HttpContext in ASP.NET Core. HttpContext is not thread-safe.

Author: ```suggestion
description: HttpContext in ASP.NET Core. HttpContext is not thread-safe and can throw NullReferenceException.
```

## HttpContext access from a background thread

> `HttpContext` isn't thread-safe. Reading or writing properties of the `HttpContext` outside of processing a request can result in a <xref:System.NullReferenceException>. In most cases, code should not access <xref:Microsoft.AspNetCore.Http.IHttpContextAccessor.HttpContext?displayProperty=nameWithType> to get data. Rather than accessing `IHttpContextAccessor`, applications should copy the data from the request on the controller or middleware and then pass it to a service.

Reviewer: I'm not sure if this is the right place for this change. The opening line from this topic suggests using `IHttpContextAccessor` with services, but then this argues against it. I think it's reasonable to reference `IHttpContextAccessor` from a service, but the complication is if said service does background work.

I wonder if this might discourage the reader from using `IHttpContextAccessor` inside services that don't perform background work, which could mean they'll end up copying data and trying to pass it through when it's not necessary to do so.

Author: Agreed.

Author: ```suggestion
`HttpContext` isn't thread-safe. Reading or writing properties of the `HttpContext` outside of processing a request can result in a <xref:System.NullReferenceException>.
```

* Copy the required data during request processing.
* Pass the copied data to a background task.

> * Do ***not*** reference `HttpContext` data in parallel tasks. The data needed from the context should be extracted before starting the parallel tasks.

Reviewer: ```suggestion
* Do ***not*** reference `HttpContext` data in parallel tasks. Extract the data needed from the context before starting the parallel tasks.
```
